Add logic to handle repeated approve request

cqbqdd11519 opened the pull request:
## Done
- Added a thread-safe map, which stores request for an Approval
- Reject the request if a request is still in progress

cqbqdd11519 committed:
[feat] Add logic to handle repeated approve request

- Added a thread-safe map, which stores request for an Approval
- Reject the request if a request is still in progress

jwkim1993 commented on pkg/server/server.go:
```diff
@@ -120,9 +122,18 @@ func handler(w http.ResponseWriter, r *http.Request) {
+	// Check if there is any ongoing request
+	_, exist := reqMap.Load(approvalName)
+	if exist {
+		respondError(w, http.StatusBadRequest, fmt.Sprintf("approval %s is still in approval/reject progress", approvalName))
```
중복된 요청에 대해서 에러를 보내는게 맞을까요?
에러를 보낸다면 400 BadRequest가 맞을까요?

jwkim1993 replied:
200 OK나 201 Accepted가 더 좋을것 같습니다. 앞선 요청을 제대로 받았고, 요청을 처리중에 있거나 처리가 완료된 상황일테니 더 좋을것같은데 혹시 어떻게 생각하시나요

cqbqdd11519 replied:
https://developer.mozilla.org/ko/docs/Web/HTTP/Status/202
202 Accepted 가 적절할것같네요

jwkim1993 commented on pkg/server/server.go:
```diff
@@ -120,9 +122,18 @@ func handler(w http.ResponseWriter, r *http.Request) {
+
 	log.Info(fmt.Sprintf("Sending request to %s", addr))
 
+	reqMap.Store(approvalName, sendReq)
```
이 코드는 동시에 Load는 막아주지만, Load를 순차적으로 했을 때 둘 다 exist 조건에 걸리지 않는다면 두 번 send 할 것 같습니다.

jwkim1993 replied:
Load후 바로 Store이니까 그럴 확률은 희박하긴 할텐데, 고민이네요

cqbqdd11519 replied:
그러네요 이부분도 레이스가 일어날수가 있어버리겠네요 지금 보니까 sync.Map에 `LoadOrStore` 함수가 있네요
코드 보니까 atomic하게 Load+Store 해주는것같으니 이거 쓰면 해결될것같습니다

jwkim1993 commented on pkg/server/server.go:
```diff
@@ -120,9 +122,18 @@ func handler(w http.ResponseWriter, r *http.Request) {
 
+	reqMap.Store(approvalName, sendReq)
 	sendResp, err := sendClient.Do(sendReq)
+	reqMap.Delete(approvalName)
```
보내고 delete를 하게되면, delete 한 후에 들어온 요청(pod이 종료되는데 수 초가 걸리기 때문에 가능성이 높음)에 대해서 처리가 불가능할것 같습니다. 혹시 pod내에 컨테이너 status를 확인해 approved, rejected를 정하는 부분에서 delete를 할 수 있을까요?

jwkim1993 replied:
그런데 그렇게 되면, 한 번 요청했다가 실제로 그 패킷이 전달이 안된 경우, 다시 못보내게 되겠네요.

jwkim1993 replied:
```go
if err != nil {
  reqMap.Delete(approvalName)
}
```
와 같이 에러가 난 경우에만 지우는건 어떨까요??

cqbqdd11519 replied:
오 고것도 괜찮은 방법일 것 같네요
성공한 경우 다시는 못보내는게 맞으니까

reqMap에 가비지가 쌓이지 않게 pod 종료될 때 성공한 approval을 지우도록 하면 되겠네요

cqbqdd11519 replied:
음 근데 생각해보니까!
리젝 사유 추가되면 Approval status를 Request 받자 마자 수정해줘야 할 것 같은데 아예 이 PR에 그 수정사항도 추가할까요?

이 경우에는
```go
_, exist := reqMap.LoadOrStore(approvalName, sendReq)
defer reqMap.Delete(approvalName)
```
이런 식으로 처리하면 괜찮지 않을까 싶어요

jwkim1993 replied:
status를 수정하고나서 함수를 빠져나갈 때 reqMap.Delete(~)한다는거죠? 완전 좋은것 같아요

cqbqdd11519 committed:
[fix] Fix possible race conditions for repeated approve request

- Used LoadOrStore instead of Load/Store respectively
- Update Approval status right after the request arrives, instead of
watching pod status
